pkg/eigenState/stakerShares/stakerShares.go

```diff
@@ -610,88 +653,172 @@ func (ss *StakerSharesModel) HandleStateChange(log *storage.TransactionLog) (int
 	return nil, nil
 }
 
+// TODO: this is not very performant in cases of large slashings and is called twice (once in CommitFinalState and once in GenerateStateRoot)
+// but prob ok?
 // prepareState prepares the state for commit by adding the new state to the existing state.
 func (ss *StakerSharesModel) prepareState(blockNumber uint64) ([]*StakerSharesDiff, error) {
-	preparedState := make([]*StakerSharesDiff, 0)
-
-	accumulatedState, ok := ss.stateAccumulator[blockNumber]
-	if !ok {
-		err := xerrors.Errorf("No accumulated state found for block %d", blockNumber)
-		ss.logger.Sugar().Errorw(err.Error(), zap.Error(err), zap.Uint64("blockNumber", blockNumber))
-		return nil, err
-	}
+	// reset the delta accumulator
+	ss.deltaAccumulator[blockNumber] = make([]*StakerShareDeltas, 0)
+	// keep track of the updated staker shares
+	updatedStakerShares := make(map[types.SlotID]*StakerSharesDiff)
 
-	slotIds := make([]types.SlotID, 0)
-	for slotId := range accumulatedState {
-		slotIds = append(slotIds, slotId)
-	}
+	for _, diff := range ss.diffAccumulator[blockNumber] {
+		switch diff.Event.(type) {
+		case *StakerShareDeltas:
+			shareDelta := diff.Event.(*StakerShareDeltas)
+			slotID := shareDelta.SlotID()
 
-	// Find only the records from the previous block, that are modified in this block
-	query := `
-		with ranked_rows as (
-			select
-				staker,
-				strategy,
-				shares,
-				block_number,
-				ROW_NUMBER() OVER (PARTITION BY staker, strategy ORDER BY block_number desc) as rn
-			from staker_shares
-			where
-				concat(staker, '_', strategy) in @slotIds
-		)
-		select
-			rr.staker,
-			rr.strategy,
-			rr.shares,
-			rr.block_number
-		from ranked_rows as rr
-		where rn = 1
-	`
-	existingRecords := make([]StakerShares, 0)
-	res := ss.DB.Model(&StakerShares{}).
-		Raw(query,
-			sql.Named("slotIds", slotIds),
-		).
-		Scan(&existingRecords)
+			// append the delta to the delta accumulator
+			ss.deltaAccumulator[blockNumber] = append(ss.deltaAccumulator[blockNumber], shareDelta)
 
-	if res.Error != nil {
-		ss.logger.Sugar().Errorw("Failed to fetch staker_shares", zap.Error(res.Error))
-		return nil, res.Error
-	}
+			currStakerShares, ok := updatedStakerShares[slotID]
+			if !ok {
+				stakerShares := &StakerShares{}
+				// get the record from the database with the given staker and strategy with the latest blockNumber
+				res := ss.DB.Model(&StakerShares{}).Where("staker = ? AND strategy = ?", shareDelta.Staker, shareDelta.Strategy).Order("block_number desc").First(&stakerShares)
```

Comment:
This DB query should be put in a function just to clean this up a bit.

`func (ss *StakerSharesModel) GetStakerSharesForStakerAndStrategy(staker, strategy string, blockNumber uint64)`

Reply:
https://github.com/Layr-Labs/sidecar/pull/68/commits/849bd0db77bd553611d6ccdf776de875882dc9c5